### DISCUSSION TITLE
Specify Library Path with analyzeHeadless?

### DISCUSSION THREAD

**Discussion by psparc82** [NationalSecurityAgency/ghidra#6955] [Q&A]
Hi,

I have an unpacked EXT4 root filesystem that I'd like to recursively analyze with the `analyzeHeadless` script. 

The libraries each binary uses are under `/lib` or `/usr/lib`, how can I tell `analyzeHeadless` to use these library paths so it can resolve the symbols and automatically apply markup to the binaries that use these symbols?

Importing each binary in the UI and pressing `Options...` (before analyzing) lets you do this but I can't figure out how to do it with the `analyzeHeadless` script

Thanks!

**Comment by ryanmkurtz** [comment]
Ghidra 11.2, which hasn't been released yet, will support a `-librarySearchPaths` headless option, which will let you do what you desire.  Are you able to build the `master` branch?

**Reply by psparc82** [reply]
Perfect! I'll give it a go

Thanks Ryan!

**Reply by ryanmkurtz** [reply]
Also, with 11.2, you won't even have to unpack the ext4 filesystem if you don't want to.

**Reply by psparc82** [reply]
I've built the master branch (v11.3) and I can see the `-librarySearchPaths` option. Do I need to have already analyzed the libraries in `/lib` and `/usr/lib` before I run `analyzeHeadless` with `-librarySearchPaths`?

In the `analyzeHeadless` output I see logs such as: `[path here] - 165 unresolved symbols, no external libraries configured - skipping (ExternalSymbolResolverAnalyzer)`

... but at this point I haven't analyzed the libraries so it makes sense `analyzeHeadless` can't use them

I'm running a test `analyzeHeadless` command that looks like this:

```
GHIDRA_HOME=/home/ghidra/ghidra_11.3_DEV ./support/analyzeHeadless /home/test_repo "test/a/b/c/ext-root" -recursive -max-cpu 10 -process -librarySearchPaths "test/a/b/c/ext-root/lib"
```

**Reply by ryanmkurtz** [reply]
`-librarySearchPaths` is for importing new libraries from disk during import.  It looks like you are using `-process` though.  Is everything already in your project?  Are these ELF files?

**Reply by psparc82** [reply]
ah that may be the issue! The binaries are already in a project but I haven't done any work to them so I can easily start a new project. These are all ELF files yes. So I need to use `-import` instead of `-process`?

**Reply by ryanmkurtz** [reply]
Yes, try something like:
```
GHIDRA_HOME=/home/ghidra/ghidra_11.3_DEV ./support/analyzeHeadless /home/test_repo test -import <path/to/file> -librarySearchPaths <path/to/libs;path/to/more/libs> -loader ElfLoader -loader-loadLibraries true
```

**Reply by ryanmkurtz** [reply]
You can also add something like 
```
-loader-libraryDestinationFolder libs
```
or something similar so your imported libraries go into a `libs/` project subfolder.

**Reply by ryanmkurtz** [reply]
If the libraries are already imported to something like `libs`, you can do 
```
-loader-linkExistingProjectLibraries true -loader-projectLibrarySearchFolder libs
```

**Reply by NoneShell** [reply]
how to set：recursive library load depth?

**Reply by ryanmkurtz** [reply]
```
-recursive <depth>
```

**Reply by psparc82** [reply]
I'm running into a few issues with `-librarySearchPaths` again. When I supply the paths everything seems to work, the `analyzeHeadless` log messages so the shared libraries being loaded for each binary but for some reason Ghidra now places all of the shared objects that each binary requires into the same folder as the binary?

For example, if I analyze a project using `analyzeHeadless` that contains a `/bin/bash` binary it copies `ld-linux-armhf.so.3` and `libc.so.6` into the `/bin` directory where `bash` is. I've added a screenshot to show this strange behavior:

![image](https://github.com/user-attachments/assets/4ba372bd-56b2-4378-953c-dbf4a7303d1f)

`libc.so.6` is from `/usr/lib/arm-linux-gnueabihf` and `ld-linux-armhf.so.3` is from `/lib`. These shared objects shouldn't be copied anywhere. Any ideas on how to prevent this from happening?

When I open `/bin/bash` in Ghidra and view the `External Programs` I can see the libraries from the `/bin` folder have been used instead of the libraries from `/lib` and `/usr/lib/arm-linux-gnueabihf`:

![image](https://github.com/user-attachments/assets/d41104b0-d52e-4405-b1c0-c869d228b765)


My `analyzeHeadless` command looks like this:

```
support/analyzeHeadless PROJECT_DIR PROJECT_NAME \
-import FILESYSTEM_DIR -recursive \
-loader ElfLoader \
-loader-loadLibraries true \
-loader-linkExistingProjectLibraries true \
-librarySearchPaths ...paths_here... 
```

Also what happens if a binary loads a library such as `libc.so.6` which is actually a symlink (on the unpacked filesystem) to the `libc-2.17.so` library? Ghidra doesn't import symlinks into the project so will it be unable to load `libc-2.17.so` for the program being analyzed?

**Reply by psparc82** [reply]
@NoneShell did you get the `-librarySearchPaths` working?

**Reply by ryanmkurtz** [reply]
`-loader-libraryDestinationFolder <project path>`

All of these are defined in `support/analyzeHeadlessREADME.html`

**Reply by psparc82** [reply]
> `-loader-libraryDestinationFolder <project path>`
> 
> All of these are defined in `support/analyzeHeadlessREADME.html`

Thanks!

if I have multiple folders with libraries in then how can I define multiple `-loader-libraryDestinationFolder` paths so my libraries don't all end up in a single folder? Can I just pass the top level project path? I'm trying to preserve the directory structure of the filesystem I've imported (I don't want the libraries to be moved/copied anywhere)

I think I'm confused as to why Ghidra is copying libraries to other folders in the project when all I'm trying to do is guide Ghidra into finding the libraries for each binary it analyzes

Also, did you see the screenshots in my previous post regarding libraries being copied into binary folders?

**Reply by ryanmkurtz** [reply]
Unfortunately we currently have no way to direct new libraries to be saved to anyplace other than the 1 folder you specify.  Would you mind making a new ticket asking for the ability to preserve the libraries directory structure on import?  I agree that would be useful.  Can you mention the symlink issue too...we need to handle that as well.

**Reply by psparc82** [reply]
> Unfortunately we currently have no way to direct new libraries to be saved to anyplace other than the 1 folder you specify. Would you mind making a new ticket asking for the ability to preserve the libraries directory structure on import? I agree that would be useful. Can you mention the symlink issue too...we need to handle that as well.

https://github.com/NationalSecurityAgency/ghidra/issues/7430

**Reply by s0kil** [reply]
Does `librarySearchPaths` resolve symlinks? `libc.so.6` -> `libc-2.25.so`?

**Reply by ryanmkurtz** [reply]
Not yet, but we are working on that.
